In [140]:
%env KERAS_BACKEND=tensorflow

env: KERAS_BACKEND=tensorflow


In [0]:
%matplotlib inline

import numpy as np
import matplotlib.pyplot as plt
from keras.datasets import mnist

(x_train, y_train), (x_test, y_test)=mnist.load_data()

x_train = x_train.reshape(60000,28,28,1)
x_test=x_test.reshape(10000,28,28,1)

from keras.utils import np_utils

y_train = np_utils.to_categorical(y_train,10)
y_test = np_utils.to_categorical(y_test,10)

#最好做normalization，不然你的CNN會很漂
x_train = x_train/255
x_test = x_test/255

##第一個設計
##Constructing CNN

In [142]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten
from keras.layers import Conv2D, MaxPool2D
from keras.optimizers import SGD

model = Sequential()

model.add(Conv2D(4,(5,5,),padding='same', input_shape=(28,28,1)))
model.add(Activation('relu'))

model.add(MaxPool2D(pool_size=(2,2)))

model.add(Conv2D(8,(5,5,),padding='same'))
model.add(Activation('relu'))

model.add(MaxPool2D(pool_size=(2,2)))

model.add(Conv2D(16,(5,5,),padding='same'))
model.add(Activation('relu'))

model.add(MaxPool2D(pool_size=(2,2)))

model.add(Flatten()) #把所有的Matrix合起來然後拉平，like ravel
model.add(Dense(9))  #Fully connected的神經元
model.add(Activation('relu'))
#輸出的時候：
model.add(Dense(10)) 

model.add(Activation('softmax')) #加起來自動要等於一

model.compile(loss='mse', optimizer=SGD(lr=0.07), metrics=['accuracy'])

model.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_37 (Conv2D)           (None, 28, 28, 4)         104       
_________________________________________________________________
activation_61 (Activation)   (None, 28, 28, 4)         0         
_________________________________________________________________
max_pooling2d_37 (MaxPooling (None, 14, 14, 4)         0         
_________________________________________________________________
conv2d_38 (Conv2D)           (None, 14, 14, 8)         808       
_________________________________________________________________
activation_62 (Activation)   (None, 14, 14, 8)         0         
_________________________________________________________________
max_pooling2d_38 (MaxPooling (None, 7, 7, 8)           0         
_________________________________________________________________
conv2d_39 (Conv2D)           (None, 7, 7, 16)          3216      
__________

In [0]:
#model.fit(x_train, y_train, batch_size=50, epochs=50)   #0.9832
#model.fit(x_train, y_train, batch_size=50, epochs=30)   #0.9791

In [144]:
#model.fit(x_train, y_train, batch_size=50, epochs=48, verbose=1, validation_data=(x_test, y_test)) #0.9840
#model.fit(x_train, y_train, batch_size=40, epochs=50)    #0.9880
model.fit(x_train, y_train, batch_size=50, epochs=48, verbose=1, validation_data=(x_test, y_test))

Train on 60000 samples, validate on 10000 samples
Epoch 1/48
60000/60000 [==============================] - 5s 92us/step - loss: 0.0899 - acc: 0.1544 - val_loss: 0.0898 - val_acc: 0.1812
Epoch 2/48
60000/60000 [==============================] - 5s 81us/step - loss: 0.0896 - acc: 0.1865 - val_loss: 0.0893 - val_acc: 0.1975
Epoch 3/48
60000/60000 [==============================] - 5s 81us/step - loss: 0.0884 - acc: 0.2058 - val_loss: 0.0868 - val_acc: 0.2550
Epoch 4/48
60000/60000 [==============================] - 5s 84us/step - loss: 0.0772 - acc: 0.3957 - val_loss: 0.0519 - val_acc: 0.6612
Epoch 5/48
60000/60000 [==============================] - 5s 90us/step - loss: 0.0291 - acc: 0.8098 - val_loss: 0.0165 - val_acc: 0.8939
Epoch 6/48
60000/60000 [==============================] - 5s 84us/step - loss: 0.0143 - acc: 0.9066 - val_loss: 0.0112 - val_acc: 0.9258
Epoch 7/48
60000/60000 [==============================] - 5s 80us/step - loss: 0.0108 - acc: 0.9293 - val_loss: 0.0087 - val_acc

In [145]:
score = model.evaluate(x_test, y_test)
print('loss:', score[0])
print('正確率', score[1])

10000/10000 [==============================] - 1s 57us/step
loss: 0.002652009207928944
正確率 0.9831


##第二個設計

這次就故意把filter設計成越來越少，看看後果是什麼

In [146]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten
from keras.layers import Conv2D, MaxPool2D
from keras.optimizers import SGD

model_1 = Sequential()

model_1.add(Conv2D(16,(5,5,),padding='same', input_shape=(28,28,1)))
model_1.add(Activation('relu'))

model_1.add(MaxPool2D(pool_size=(2,2)))

model_1.add(Conv2D(8,(5,5,),padding='same'))
model_1.add(Activation('relu'))

model_1.add(MaxPool2D(pool_size=(2,2)))

model_1.add(Conv2D(4,(5,5,),padding='same'))
model_1.add(Activation('relu'))

model_1.add(MaxPool2D(pool_size=(2,2)))

model_1.add(Flatten()) #把所有的Matrix合起來然後拉平，like ravel
model_1.add(Dense(9))  #Fully connected的神經元
model_1.add(Activation('relu'))
#輸出的時候：
model_1.add(Dense(10)) 

model_1.add(Activation('softmax')) #加起來自動要等於一

model_1.compile(loss='mse', optimizer=SGD(lr=0.07), metrics=['accuracy'])

model_1.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_40 (Conv2D)           (None, 28, 28, 16)        416       
_________________________________________________________________
activation_66 (Activation)   (None, 28, 28, 16)        0         
_________________________________________________________________
max_pooling2d_40 (MaxPooling (None, 14, 14, 16)        0         
_________________________________________________________________
conv2d_41 (Conv2D)           (None, 14, 14, 8)         3208      
_________________________________________________________________
activation_67 (Activation)   (None, 14, 14, 8)         0         
_________________________________________________________________
max_pooling2d_41 (MaxPooling (None, 7, 7, 8)           0         
_________________________________________________________________
conv2d_42 (Conv2D)           (None, 7, 7, 4)           804       
__________

In [147]:
model_1.fit(x_train, y_train, batch_size=50, epochs=48, verbose=1, validation_data=(x_test, y_test)) #0.9758

Train on 60000 samples, validate on 10000 samples
Epoch 1/48
60000/60000 [==============================] - 6s 95us/step - loss: 0.0896 - acc: 0.1309 - val_loss: 0.0890 - val_acc: 0.1485
Epoch 2/48
60000/60000 [==============================] - 5s 84us/step - loss: 0.0864 - acc: 0.2215 - val_loss: 0.0787 - val_acc: 0.3861
Epoch 3/48
60000/60000 [==============================] - 5s 84us/step - loss: 0.0569 - acc: 0.5864 - val_loss: 0.0370 - val_acc: 0.7538
Epoch 4/48
60000/60000 [==============================] - 5s 84us/step - loss: 0.0335 - acc: 0.7721 - val_loss: 0.0293 - val_acc: 0.7993
Epoch 5/48
60000/60000 [==============================] - 5s 83us/step - loss: 0.0286 - acc: 0.8048 - val_loss: 0.0258 - val_acc: 0.8233
Epoch 6/48
60000/60000 [==============================] - 5s 83us/step - loss: 0.0261 - acc: 0.8218 - val_loss: 0.0258 - val_acc: 0.8261
Epoch 7/48
60000/60000 [==============================] - 5s 83us/step - loss: 0.0240 - acc: 0.8334 - val_loss: 0.0202 - val_acc

In [148]:
score = model_1.evaluate(x_test, y_test)
print('loss:', score[0])
print('正確率', score[1])

10000/10000 [==============================] - 1s 56us/step
loss: 0.004193800970243501
正確率 0.9733


##第三個設計
就是不要做normalization

In [0]:
%matplotlib inline

import numpy as np
import matplotlib.pyplot as plt
from keras.datasets import mnist

(x_train_1, y_train_1), (x_test_1, y_test_1)=mnist.load_data()

x_train_1 = x_train_1.reshape(60000,28,28,1)
x_test_1=x_test_1.reshape(10000,28,28,1)

from keras.utils import np_utils

y_train_1 = np_utils.to_categorical(y_train_1,10)
y_test_1 = np_utils.to_categorical(y_test_1,10)


In [150]:
from keras.models import Sequential
from keras.layers import Dense, Activation, Flatten
from keras.layers import Conv2D, MaxPool2D
from keras.optimizers import SGD

model_2 = Sequential()

model_2.add(Conv2D(4,(5,5,),padding='same', input_shape=(28,28,1)))
model_2.add(Activation('relu'))

model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Conv2D(8,(5,5,),padding='same'))
model_2.add(Activation('relu'))

model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Conv2D(16,(5,5,),padding='same'))
model_2.add(Activation('relu'))

model_2.add(MaxPool2D(pool_size=(2,2)))

model_2.add(Flatten()) #把所有的Matrix合起來然後拉平，like ravel
model_2.add(Dense(9))  #Fully connected的神經元
model_2.add(Activation('relu'))
#輸出的時候：
model_2.add(Dense(10)) 

model_2.add(Activation('softmax')) #加起來自動要等於一

model_2.compile(loss='mse', optimizer=SGD(lr=0.07), metrics=['accuracy'])

model_2.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_43 (Conv2D)           (None, 28, 28, 4)         104       
_________________________________________________________________
activation_71 (Activation)   (None, 28, 28, 4)         0         
_________________________________________________________________
max_pooling2d_43 (MaxPooling (None, 14, 14, 4)         0         
_________________________________________________________________
conv2d_44 (Conv2D)           (None, 14, 14, 8)         808       
_________________________________________________________________
activation_72 (Activation)   (None, 14, 14, 8)         0         
_________________________________________________________________
max_pooling2d_44 (MaxPooling (None, 7, 7, 8)           0         
_________________________________________________________________
conv2d_45 (Conv2D)           (None, 7, 7, 16)          3216      
__________

In [151]:
model_2.fit(x_train_1, y_train_1, batch_size=50, epochs=48, verbose=1, validation_data=(x_test_1, y_test_1)) #0.9758

Train on 60000 samples, validate on 10000 samples
Epoch 1/48
60000/60000 [==============================] - 5s 88us/step - loss: 0.0769 - acc: 0.2899 - val_loss: 0.0640 - val_acc: 0.4016
Epoch 2/48
60000/60000 [==============================] - 5s 78us/step - loss: 0.0582 - acc: 0.4910 - val_loss: 0.0502 - val_acc: 0.5642
Epoch 3/48
60000/60000 [==============================] - 5s 86us/step - loss: 0.0499 - acc: 0.5682 - val_loss: 0.0493 - val_acc: 0.5735
Epoch 4/48
60000/60000 [==============================] - 5s 84us/step - loss: 0.0480 - acc: 0.5766 - val_loss: 0.0468 - val_acc: 0.5813
Epoch 5/48
60000/60000 [==============================] - 5s 78us/step - loss: 0.0469 - acc: 0.5830 - val_loss: 0.0460 - val_acc: 0.5853
Epoch 6/48
60000/60000 [==============================] - 5s 78us/step - loss: 0.0449 - acc: 0.6052 - val_loss: 0.0395 - val_acc: 0.6661
Epoch 7/48
60000/60000 [==============================] - 5s 78us/step - loss: 0.0376 - acc: 0.7065 - val_loss: 0.0346 - val_acc

In [152]:
score = model_2.evaluate(x_test_1, y_test_1)
print('loss:', score[0])
print('正確率', score[1])

10000/10000 [==============================] - 1s 59us/step
loss: 0.004541242292661053
正確率 0.9747


##結論：
### 第一個CNN：只改了epoch和batch_size以提高正確率
正確率：0.9831

### 第二個CNN：把Filter設計成越來越少，16、8、4，其他保持一樣
正確率：0.9733

### 第三個CNN：不要normalization，其他保持一樣
正確率：0.9747